tests/end_to_end_test/snapshot_fail_test.go

```diff
@@ -0,0 +1,140 @@
+package endtoend_test
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"testing"
+
+	"github.com/kopia/kopia/tests/testenv"
+)
+
+func TestSnapshotFail(t *testing.T) {
+	t.Parallel()
+
+	e := testenv.NewCLITest(t)
+	defer e.Cleanup(t)
+	defer e.RunAndExpectSuccess(t, "repo", "disconnect")
+
+	e.RunAndExpectSuccess(t, "repo", "create", "filesystem", "--path", e.RepoDir)
+	e.RunAndExpectSuccess(t, "policy", "set", "--global", "--keep-latest", strconv.Itoa(1<<31-1))
+
+	parentDirOfSource := makeScratchDir(t)
+	source := filepath.Join(parentDirOfSource, "source")
+
+	// Test snapshot of nonexistent directory fails
+	e.RunAndExpectFailure(t, "snapshot", "create", source)
+
+	testenv.MustCreateDirectoryTree(t, source, testenv.DirectoryTreeOptions{
+		Depth:                  2,
+		MaxSubdirsPerDirectory: 1,
+		MaxFilesPerDirectory:   1,
+	})
+
+	// Create snapshot
+	e.RunAndExpectSuccess(t, "snapshot", "create", source)
+
+	numSuccessfulSnapshots := 1
+
+	// Test the root dir permissions
+	for _, tc := range []struct {
+		desc      string
+		parentDir string
+	}{
+		{
+			desc:      "Modify permissions of the snapshot root directory",
+			parentDir: filepath.Dir(source),
+		},
+		{
+			desc:      "Modify permissions of entries in the snapshot root dir",
+			parentDir: source,
+		},
+		{
+			desc:      "Modify permissions of entries in a subdir of the snapshot root",
+			parentDir: findASubDirFilePath(t, source),
+		},
+	} {
+		t.Log(tc.desc)
+		numSuccessfulSnapshots += testPermissions(e, t, source, tc.parentDir)
+	}
+
+	// check the number of snapshots that succeeded match the length of
+	// a snap list output
+	si := e.ListSnapshotsAndExpectSuccess(t)
+	if got, want := len(si), 1; got != want {
+		t.Fatalf("got %v sources, wanted %v", got, want)
+	}
+
+	if got, want := len(si[0].Snapshots), numSuccessfulSnapshots; got != want {
+		t.Fatalf("got %v snapshots, wanted %v", got, want)
+	}
+}
+
+func findASubDirFilePath(t *testing.T, parent string) string {
+	fileInfoList, err := ioutil.ReadDir(parent)
+	testenv.AssertNoError(t, err)
+
+	for _, fi := range fileInfoList {
+		if fi.IsDir() {
+			return filepath.Join(parent, fi.Name())
+		}
+	}
+
+	t.Fatalf("Could not find a subdirectory in parent dir %v", parent)
+	return ""
+}
+
+// Perm constants
+const (
+	execOffset     = 0
+	readOffset     = 2
+	userPermOffset = 6
+)
+
+// testPermissions iterates over readable and executable permission states, testing
+// files and directories (if present). It issues the kopia snapshot command
+// against "source" and will test permissions against all entries in "parentDir".
+// It returns the number of successful snapshot operations.
+func testPermissions(e *testenv.CLITest, t *testing.T, source, parentDir string) int {
```

Comment:
In the context of the comment above:

* the "expected success" value should be passed, and then used as part of the check below (along with the error permissions check), instead of checking whether or not it is a directory.
*  renaming the parameters may help with clarity.

Something like:

```suggestion
func testPermissions(e *testenv.CLITest, t *testing.T, snapSource, pathToModify string) int {
```